src/NuGet.Core/NuGet.Build.Tasks/GetGlobalPropertyValueTask.cs

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Build.Framework;
+using Newtonsoft.Json;
+using Task = Microsoft.Build.Utilities.Task;
+
+
+namespace NuGet.Build.Tasks
+{
+    public class GetGlobalPropertyValueTask : Task
+    {
+        [Required]
+        public string PropertyName { get; set; }
+
+        [Output]
+        public string GlobalPropertyValue { get; set; }
+
+        [Output]
+        public bool CheckCompleted { get; set; }
+
+        public override bool Execute()
+        {
+            var globalProperties = GetGlobalProperties();
+
+            if (globalProperties != null)
+            {
+                var dictionaryWithOrdinal = globalProperties.ToDictionary(i => i.Key, i => i.Value, StringComparer.OrdinalIgnoreCase);
+                CheckCompleted = true;
+                if (dictionaryWithOrdinal.TryGetValue(PropertyName, out string globalProperty))
+                {
+                    GlobalPropertyValue = globalProperty;
+                }
+            }
+            return !Log.HasLoggedErrors;
+        }
+
+        /// <summary>
+        /// Get the global property from the IBuildEngine API. 
+        /// </summary>
+        /// <returns>Returns the dictionary with the global properties if they can be accessed. <see langword="null"/> otherwise, which means that the msbuild version doesn't implement this API. </returns>
+        internal IReadOnlyDictionary<string, string> GetGlobalProperties()
+        {
+#if IS_CORECLR
+            // MSBuild 16.5 and above has a method to get the global properties, older versions do not
+            IReadOnlyDictionary<string, string> msBuildGlobalProperties = BuildEngine is IBuildEngine6 buildEngine6
+                ? buildEngine6.GetGlobalProperties()
+                : null;
+#else
+            IReadOnlyDictionary<string, string> msBuildGlobalProperties = null;
+
+            // MSBuild 16.5 added a new interface, IBuildEngine6, which has a GetGlobalProperties() method.  However, we compile against
+            // Microsoft.Build.Framework version 4.0 when targeting .NET Framework, so reflection is required since type checking
+            // can't be done at compile time
+            var buildEngine6Type = typeof(IBuildEngine).Assembly.GetType("Microsoft.Build.Framework.IBuildEngine6");
+
+            if (buildEngine6Type != null)
+            {
+                var getGlobalPropertiesMethod = buildEngine6Type.GetMethod("GetGlobalProperties", BindingFlags.Instance | BindingFlags.Public);
```

Comment:
Maybe you don't need to lookup `IBuildEngine6`, it seems like it is a regular method:

https://github.com/dotnet/msbuild/blob/235c0e7462df92e0bada1a5a9f6dca38aeba6a38/src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs#L678

Maybe something like this would work, and use a little less System.Reflection:

```csharp
var getGlobalPropertiesMethod = BuildEngine.GetType().GetMethod("GetGlobalProperties", BindingFlags.Instance | BindingFlags.Public);
```

Reply:
Thanks, I'll try that out.

Reply:
Works.